sdks/java/core/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataGrpcMultiplexer.java

```diff
@@ -104,12 +131,43 @@ private CompletableFuture<CloseableFnDataReceiver<BeamFnApi.Elements>> receiverF
    */
   public void registerConsumer(
       String instructionId, CloseableFnDataReceiver<BeamFnApi.Elements> receiver) {
-    receiverFuture(instructionId).complete(receiver);
+    @Nullable
+    CompletableFuture<CloseableFnDataReceiver<BeamFnApi.Elements>> receiverFuture =
+        receiverFuture(instructionId);
+    if (receiverFuture == null) {
+      throw new IllegalArgumentException("Instruction id was poisoned");
+    } else {
+      receiverFuture.complete(receiver);
+    }
   }
 
   /** Unregisters a consumer. */
   public void unregisterConsumer(String instructionId) {
-    receivers.remove(instructionId);
+    @Nullable
+    CompletableFuture<CloseableFnDataReceiver<BeamFnApi.Elements>> receiverFuture =
+        receivers.remove(instructionId);
+    if (receiverFuture != null && !receiverFuture.isDone()) {
+      throw new IllegalStateException("Unregistering consumer which was not registered.");
```

Comment:
This error message doesn't seem to match the clauses of the if statement.

Reply:
Added explanation comment.